client/tso_service_discovery.go

```diff
@@ -0,0 +1,306 @@
+func (c *tsoServiceDiscovery) GetServingAddr() string {
+	return c.getPrimaryAddr()
+}
+
+// GetBackupAddrs gets the addresses of the current reachable and healthy
+// backup service endpoints randomly. Backup service endpoints are secondaries in
+// a primary/secondary configured cluster.
+func (c *tsoServiceDiscovery) GetBackupAddrs() []string {
+	return c.getSecondaryAddrs()
+}
+
+// GetOrCreateGRPCConn returns the corresponding grpc client connection of the given addr.
+func (c *tsoServiceDiscovery) GetOrCreateGRPCConn(addr string) (*grpc.ClientConn, error) {
+	return grpcutil.GetOrCreateGRPCConn(c.ctx, &c.clientConns, addr, c.tlsCfg, c.option.gRPCDialOptions...)
+}
+
+// ScheduleCheckMemberChanged is used to trigger a check to see if there is any change in ervice endpoints.
+func (c *tsoServiceDiscovery) ScheduleCheckMemberChanged() {
+	select {
+	case c.checkMembershipCh <- struct{}{}:
+	default:
+	}
+}
+
+// Immediately check if there is any membership change among the primary/secondaries in
+// a primary/secondary configured cluster.
+func (c *tsoServiceDiscovery) CheckMemberChanged() error {
+	return c.updateMember()
+}
+
+// AddServingAddrSwitchedCallback adds callbacks which will be called when the primary in
+// a primary/secondary configured cluster is switched.
+func (c *tsoServiceDiscovery) AddServingAddrSwitchedCallback(callbacks ...func()) {
+	c.primarySwitchedCbs = append(c.primarySwitchedCbs, callbacks...)
+}
+
+// AddServiceAddrsSwitchedCallback adds callbacks which will be called when any primary/secondary
+// in a primary/secondary configured cluster is changed.
+func (c *tsoServiceDiscovery) AddServiceAddrsSwitchedCallback(callbacks ...func()) {
+	c.membersChangedCbs = append(c.membersChangedCbs, callbacks...)
+}
+
+// AddTSOServAddrsUpdatedCallback adds callbacks which will be called
+// when the global/local tso allocator leader list is updated.
+func (c *tsoServiceDiscovery) AddTSOServAddrsUpdatedCallback(callbacks ...tsoServAddrsUpdatedFunc) {
+	c.allocPrimariesUpdatedCbs = append(c.allocPrimariesUpdatedCbs, callbacks...)
+}
+
+// getPrimaryAddr returns the primary address.
+func (c *tsoServiceDiscovery) getPrimaryAddr() string {
+	primaryAddr := c.primary.Load()
+	if primaryAddr == nil {
+		return ""
+	}
+	return primaryAddr.(string)
+}
+
+// getSecondaryAddrs returns the secondary addresses.
+func (c *tsoServiceDiscovery) getSecondaryAddrs() []string {
+	secondaryAddrs := c.secondaries.Load()
+	if secondaryAddrs == nil {
+		return []string{}
+	}
+	return secondaryAddrs.([]string)
+}
+
+func (c *tsoServiceDiscovery) switchPrimary(addrs []string) error {
+	// FIXME: How to safely compare primary urls? For now, only allows one client url.
+	addr := addrs[0]
+	oldPrimary := c.getPrimaryAddr()
+	if addr == oldPrimary {
+		return nil
+	}
+
+	if _, err := c.GetOrCreateGRPCConn(addr); err != nil {
+		log.Warn("[pd(tso)] failed to connect primary", zap.String("primary", addr), errs.ZapError(err))
```

Comment:
How about changing all `[pd(tso)]` to `[tso]`?

Reply:
And we also have `[pd/tso]`. We need to unify them.

Reply:
changed.